server/src/main/java/org/elasticsearch/cluster/coordination/JoinTaskExecutor.java

```diff
@@ -41,19 +41,28 @@
 
     private final AllocationService allocationService;
     private final RerouteService rerouteService;
-    private final long term;
 
-    public JoinTaskExecutor(AllocationService allocationService, RerouteService rerouteService, long term) {
+    public JoinTaskExecutor(AllocationService allocationService, RerouteService rerouteService) {
         this.allocationService = allocationService;
         this.rerouteService = rerouteService;
-        this.term = term;
     }
 
     @Override
     public ClusterState execute(ClusterState currentState, List<TaskContext<JoinTask>> joinTaskContexts) throws Exception {
         // The current state that MasterService uses might have been updated by a (different) master in a higher term already. If so, stop
         // processing the current cluster state update, there's no point in continuing to compute it as it will later be rejected by
         // Coordinator#publish anyhow.
+        assert joinTaskContexts.isEmpty() == false : "Expected to have non empty join tasks list";
+
+        var term = joinTaskContexts.stream().mapToLong(t -> t.getTask().term()).max().getAsLong();
+
+        var split = joinTaskContexts.stream().collect(Collectors.partitioningBy(t -> t.getTask().term() == term));
+        for (TaskContext<JoinTask> outdated : split.get(false)) {
+            outdated.onFailure(new NotMasterException("Join task term [" + term + "] is outdated"));
+        }
+
+        joinTaskContexts = split.get(true);// process up-to-date tasks
```

Comment:
Nit: I'm not convinced this comment makes things any clearer. Are you sure it's useful?

```suggestion
        joinTaskContexts = split.get(true);
```